### PR TITLE
std: add io.MultiOutStream

### PR DESCRIPTION
**Pull request by daurnimator**
Requested by @Sobeston

### CODE REVIEW

**Comment by daurnimator** [comment]
On @haze 's suggestion I attempted to do an async solution:
```diff
diff --git a/lib/std/io/multi_out_stream.zig b/lib/std/io/multi_out_stream.zig
index 2197defbb..ae606df04 100644
--- a/lib/std/io/multi_out_stream.zig
+++ b/lib/std/io/multi_out_stream.zig
@@ -22,18 +22,14 @@ pub fn MultiOutStream(comptime OutStreams: type) type {
         }
 
         pub fn write(self: *Self, bytes: []const u8) Error!usize {
-            if (comptime self.streams.len == 0) return bytes.len;
-
-            // only first stream is allowed to do a partial write
-            // all subsequent streams do a `.writeAll`
-            const bytes_to_write = try self.streams[0].write(bytes);
-            const slice_to_write = bytes[0..bytes_to_write];
-            comptime var i = 1;
+            var batch = std.event.Batch(error{}!void, self.streams.len, .auto_async).init();
+            comptime var i = 0;
             inline while (i < self.streams.len) : (i += 1) {
                 const stream = self.streams[i];
-                try stream.writeAll(slice_to_write);
+                batch.add(&async stream.writeAll(bytes));
             }
-            return bytes_to_write;
+            try batch.wait();
+            return bytes.len;
         }
     };
 }
```

but this failed with:
```
$ zig test lib/std/io/multi_out_stream.zig 
Semantic Analysis [697/1089] ./lib/std/io/multi_out_stream.zig:29:28: error: expected type 'anyframe->batch!void', found '*@Frame(std.io.out_stream.OutStream(*std.io.fixed_buffer_stream.FixedBufferStream([]u8),std.io.fixed_buffer_stream.WriteError,std.io.fixed_buffer_stream.FixedBufferStream([]u8).write).writeAll)'
                batch.add(&async stream.writeAll(bytes));
                           ^
```

I can't seem to figure out the right cast/coercion to make the cast to `anyframe` work.

File: lib/std/io/multi_out_stream.zig [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+            // only first stream is allowed to do a partial write
+            // all subsequent streams do a `.writeAll`
+            const bytes_to_write = try self.streams[0].write(bytes);
+            const slice_to_write = bytes[0..bytes_to_write];
+            comptime var i = 1;
+            inline while (i < self.streams.len) : (i += 1) {
+                const stream = self.streams[i];
+                try stream.writeAll(slice_to_write);
+            }
+            return bytes_to_write;
```
Comment (data-man) [inline lib/std/io/multi_out_stream.zig:36]: Maybe summarize all bytes written?

Comment (daurnimator) [reply lib/std/io/multi_out_stream.zig:36]: no.... that's not what the return value of an out stream's `write` means.

Comment (data-man) [reply lib/std/io/multi_out_stream.zig:36]: Why no?
Suppose you wrote 1 byte into stream1 and 1 byte into stream2. Obviously, you wrote 2 bytes.
E.g. you'll get the wrong result for `CountingOutStream` with `MultiOutStream`.

Comment (daurnimator) [reply lib/std/io/multi_out_stream.zig:36]: The purpose of the return value is to know how many bytes of the input argument were written out, and can hence be discarded.

**Pull request merged**